JustSaying/Messaging/MessageProcessingStrategies/Throttled.cs

```diff
@@ -3,81 +3,157 @@
+        /// </summary>
+        ~Throttled()
+        {
+            Dispose(false);
+        }
+
+        /// <inheritdoc />
+        public int AvailableWorkers => _semaphore.CurrentCount;
+
+        /// <inheritdoc />
+        public int MaxWorkers { get; }
+
+        /// <summary>
+        /// Gets the timeout to use for starting a worker.
+        /// </summary>
+        public TimeSpan StartTimeout { get; }
+
+        /// <inheritdoc />
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        /// <inheritdoc />
+        public virtual async Task<bool> StartWorkerAsync(Func<Task> action, CancellationToken cancellationToken)
         {
             try
             {
-                await action().ConfigureAwait(false);
+                if (await _semaphore.WaitAsync(StartTimeout, cancellationToken).ConfigureAwait(false))
+                {
+                    try
+                    {
+                        _ = Task.Factory.StartNew(async (object semaphore) =>
+                        {
+                            try
+                            {
+                                await action().ConfigureAwait(false);
+                            }
+                            finally
+                            {
+                                ((SemaphoreSlim)semaphore).Release();
+                            }
+                        },
+                        _semaphore,
+                        CancellationToken.None,
+                        TaskCreationOptions.None,
+                        TaskScheduler.Default);
+                    }
+                    catch (Exception)
+                    {
+                        _semaphore.Release();
+                        throw;
+                    }
+
+                    return true;
+                }
+                else
+                {
+                    _messageMonitor.IncrementThrottlingStatistic();
+                    _messageMonitor.HandleThrottlingTime(StartTimeout);
+
+                    _logger.LogWarning("Message handling was throttled waiting for the semaphore for {StartTimeout}.", StartTimeout);
+
+                    return false;
+                }
             }
-            finally
+            catch (OperationCanceledException)
             {
-                _semaphore.Release();
+                return false;
             }
         }
 
-
-        public async Task WaitForAvailableWorkers()
+        /// <inheritdoc />
+        public async Task<int> WaitForAvailableWorkerAsync()
         {
-            if (_semaphore.CurrentCount != 0)
-                return;
-
-            _messageMonitor.IncrementThrottlingStatistic();
+            if (_semaphore.CurrentCount == 0)
+            {
+                _messageMonitor.IncrementThrottlingStatistic();
 
-            var watch = Stopwatch.StartNew();
-            await AsTask(_semaphore.AvailableWaitHandle).ConfigureAwait(false);
```

Comment:
Ironically you only need to dispose a `SemphoreSlim` when you access the `AvailableWaitHandle` property, but `IDisposable` is a welcome addition here anyway.